File: android/src/main/java/com/avinashbhalki/capacitor/twilio/video/TwilioVideoPlugin.kt
package com.avinashbhalki.capacitor.twilio.video

import android.Manifest
import android.content.Intent
import android.util.Log
import com.getcapacitor.JSArray
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import com.getcapacitor.annotation.Permission
import com.twilio.video.Room

@CapacitorPlugin(
    name = "TwilioVideo",
    permissions = [
        Permission(strings = [Manifest.permission.CAMERA], alias = "camera"),
        Permission(strings = [Manifest.permission.RECORD_AUDIO], alias = "audio")
    ]
)
class TwilioVideoPlugin : Plugin() {

    companion object {
        private var pluginInstance: TwilioVideoPlugin? = null

        fun getInstance(): TwilioVideoPlugin? {
            return pluginInstance
        }
    }

    init {
        pluginInstance = this
    }

    @PluginMethod
    fun joinRoom(call: PluginCall) {
        val token = call.getString("token")
        val roomName = call.getString("roomName")
        val roomId = call.getString("roomId")
        val identity = call.getString("identity")
        val role = call.getString("role")
        val tenantId = call.getInt("tenantId")

        if (token == null || token.isEmpty()) {
            call.reject("Token is required")
            return
        }

        if ((roomName == null || roomName.isEmpty()) && (roomId == null || roomId.isEmpty())) {
            call.reject("Either roomName or roomId must be provided")
            return
        }

        // Validate role if provided
        if (role != null && !listOf("mht", "cct", "patient").contains(role)) {
            call.reject("Invalid role. Allowed values: mht, cct, patient")
            return
        }

        val intent = Intent(activity, VideoCallActivity::class.java)
        intent.putExtra("token", token)
        intent.putExtra("roomName", roomName ?: roomId)
        intent.putExtra("identity", identity)
        intent.putExtra("role", role)
        if (tenantId != null) {
            intent.putExtra("tenantId", tenantId)
        }

        activity.startActivity(intent)
        call.resolve()
    }

    @PluginMethod
    fun leaveRoom(call: PluginCall) {
        VideoCallActivity.getInstance()?.disconnect()
        call.resolve()
    }

    @PluginMethod
    fun muteAudio(call: PluginCall) {
        val muted = call.getBoolean("muted", false) ?: false
        VideoCallActivity.getInstance()?.muteAudio(muted)
        call.resolve()
    }

    @PluginMethod
    fun enableVideo(call: PluginCall) {
        val enabled = call.getBoolean("enabled", true) ?: true
        VideoCallActivity.getInstance()?.enableVideo(enabled)
        call.resolve()
    }

    @PluginMethod
    fun flipCamera(call: PluginCall) {
        VideoCallActivity.getInstance()?.flipCamera()
        call.resolve()
    }

    @PluginMethod
    fun setSpeaker(call: PluginCall) {
        val enabled = call.getBoolean("enabled", true) ?: true
        VideoCallActivity.getInstance()?.setSpeaker(enabled)
        call.resolve()
    }

    @PluginMethod
    fun sendUsersList(call: PluginCall) {
        Log.d("TwilioVideoPlugin", "sendUsersList method invoked")

        val selectedRoleKey = call.getString("selectedRoleKey")
        val usersArray = call.getArray("users")

        if (selectedRoleKey == null || selectedRoleKey.isEmpty()) {
            Log.e("TwilioVideoPlugin", "sendUsersList failed: selectedRoleKey is required")
            call.reject("selectedRoleKey is required")
            return
        }

        if (!listOf("mht", "cct", "patient", "participant").contains(selectedRoleKey)) {
            Log.e("TwilioVideoPlugin", "sendUsersList failed: Invalid selectedRoleKey: $selectedRoleKey")
            call.reject("Invalid selectedRoleKey. Allowed values: mht, cct, patient, participant")
            return
        }

        val users = mutableListOf<Map<String, String>>()
        if (usersArray != null) {
            Log.d("TwilioVideoPlugin", "sendUsersList parsing ${usersArray.length()} users")
            for (i in 0 until usersArray.length()) {
                try {
                    val userObj = usersArray.getJSONObject(i)
                    val id = userObj.optString("id", "")
                    val userId = userObj.optString("user_id", "")
                    val fullName = userObj.optString("full_name", "")

                    if (id.isNotEmpty() && userId.isNotEmpty() && fullName.isNotEmpty()) {
                        Log.d("TwilioVideoPlugin", "sendUsersList parsed user: $fullName")
                        users.add(mapOf(
                            "id" to id,
                            "user_id" to userId,
                            "full_name" to fullName
                        ))
                    } else {
                        Log.w("TwilioVideoPlugin", "sendUsersList skipping incomplete user at index $i")
                    }
                } catch (e: Exception) {
                    Log.e("TwilioVideoPlugin", "sendUsersList error parsing user at index $i: ${e.message}")
                }
            }
        } else {
            Log.d("TwilioVideoPlugin", "sendUsersList received null users array")
        }

        Log.i("TwilioVideoPlugin", "sendUsersList calling handleUsersList with ${users.size} users for role: $selectedRoleKey")
        VideoCallActivity.getInstance()?.handleUsersList(selectedRoleKey, users)
        call.resolve()
    }

    @PluginMethod
    fun sendFormsList(call: PluginCall) {
        Log.d("TwilioVideoPlugin", "sendFormsList method invoked")

        val formsArray = call.getArray("forms")
        val forms = mutableListOf<Map<String, Any>>()

        if (formsArray != null) {
            Log.d("TwilioVideoPlugin", "sendFormsList parsing ${formsArray.length()} forms")
            for (i in 0 until formsArray.length()) {
                try {
                    val formObj = formsArray.getJSONObject(i)
                    val formMap = mutableMapOf<String, Any>()

                    formObj.keys().forEach { key ->
                        val value = formObj.get(key)
                        formMap[key] = value
                    }

                    forms.add(formMap)
                    Log.d("TwilioVideoPlugin", "sendFormsList parsed form: ${formObj.optString("name", "Unknown")}")
                } catch (e: Exception) {
                    Log.e("TwilioVideoPlugin", "sendFormsList error parsing form at index $i: ${e.message}")
                }
            }
        } else {
            Log.d("TwilioVideoPlugin", "sendFormsList received null forms array")
        }

        Log.i("TwilioVideoPlugin", "sendFormsList calling handleFormsList with ${forms.size} forms")
        VideoCallActivity.getInstance()?.handleFormsList(forms)
        call.resolve()
    }

    fun notifyRoomConnected(roomName: String) {
        val data = JSObject()
        data.put("roomName", roomName)
        notifyListeners("roomConnected", data)
    }

    fun notifyRoomDisconnected(room: com.twilio.video.Room) {
        // Build room object structure
        val roomData = JSObject()
        roomData.put("sid", room.sid)
        roomData.put("name", room.name)
        roomData.put("state", roomStateToString(room.state))

        // Add local participant
        room.localParticipant?.let { localParticipant ->
            val localParticipantData = JSObject()
            localParticipantData.put("sid", localParticipant.sid)
            localParticipantData.put("identity", localParticipant.identity)
            roomData.put("localParticipant", localParticipantData)
        }

        // Add remote participants
        val remoteParticipants = mutableListOf<JSObject>()
        room.remoteParticipants.forEach { participant ->
            val participantData = JSObject()
            participantData.put("sid", participant.sid)
            participantData.put("identity", participant.identity)
            remoteParticipants.add(participantData)
        }
        roomData.put("remoteParticipants", remoteParticipants)

        val data = JSObject()
        data.put("room", roomData)
        notifyListeners("roomDisconnected", data)
    }

    private fun roomStateToString(state: com.twilio.video.Room.State): String {
        return when (state) {
            com.twilio.video.Room.State.CONNECTING -> "connecting"
            com.twilio.video.Room.State.CONNECTED -> "connected"
            com.twilio.video.Room.State.RECONNECTING -> "reconnecting"
            com.twilio.video.Room.State.DISCONNECTED -> "disconnected"
        }
    }

    fun notifyParticipantJoined(identity: String) {
        val data = JSObject()
        data.put("identity", identity)
        notifyListeners("participantJoined", data)
    }

    fun notifyParticipantLeft(identity: String) {
        val data = JSObject()
        data.put("identity", identity)
        notifyListeners("participantLeft", data)
    }

    fun notifyNetworkQualityChanged(identity: String, level: Int, isLocal: Boolean) {
        val data = JSObject()
        data.put("identity", identity)
        data.put("level", level)
        data.put("isLocal", isLocal)
        notifyListeners("networkQualityChanged", data)
    }

    fun notifyDominantSpeakerChanged(identity: String?) {
        val data = JSObject()
        data.put("identity", identity)
        notifyListeners("dominantSpeakerChanged", data)
    }

    fun notifyRoomAutoClosed(reason: String) {
        val data = JSObject()
        data.put("reason", reason)
        notifyListeners("roomAutoClosed", data)
    }

    fun notifyRoomError(code: String, message: String, isFatal: Boolean) {
        val data = JSObject()
        data.put("code", code)
        data.put("message", message)
        data.put("isFatal", isFatal)
        notifyListeners("roomError", data)
    }

    fun notifyRoleSelected(selectedRoleKey: String, identity: String?, role: String?, tenantId: Int?, roomName: String?, roomSID: String?, secondParticipantRole: String?, secondParticipantIdentity: String?) {
        val data = JSObject()
        data.put("selectedRoleKey", selectedRoleKey)
        if (identity != null) data.put("identity", identity)
        if (role != null) data.put("role", role)
        if (tenantId != null) data.put("tenantId", tenantId)
        if (roomName != null) data.put("roomName", roomName)
        if (roomSID != null) data.put("roomSID", roomSID)
        if (secondParticipantRole != null) data.put("secondParticipantRole", secondParticipantRole)
        if (secondParticipantIdentity != null) data.put("secondParticipantIdentity", secondParticipantIdentity)
        notifyListeners("roleSelected", data)
    }

    fun notifyUsersListLoaded(selectedRoleKey: String, userCount: Int) {
        val data = JSObject()
        data.put("selectedRoleKey", selectedRoleKey)
        data.put("userCount", userCount)
        notifyListeners("usersListLoaded", data)
    }

    fun notifyUserSelected(id: String, userId: String, fullName: String, selectedRoleKey: String, tenantId: Int?, role: String?, roomName: String?, roomSID: String?) {
        val data = JSObject()
        data.put("id", id)
        data.put("user_id", userId)
        data.put("full_name", fullName)
        data.put("selectedRoleKey", selectedRoleKey)
        if (tenantId != null) data.put("tenantId", tenantId)
        if (role != null) data.put("role", role)
        if (roomName != null) data.put("roomName", roomName)
        if (roomSID != null) data.put("roomSID", roomSID)
        notifyListeners("userSelected", data)
    }

    fun notifyPopupDismissed(popupType: String, reason: String) {
        val data = JSObject()
        data.put("popupType", popupType)
        data.put("reason", reason)
        notifyListeners("popupDismissed", data)
    }

    fun notifyPopupError(message: String, popupType: String) {
        val data = JSObject()
        data.put("message", message)
        data.put("popupType", popupType)
        notifyListeners("popupError", data)
    }

    fun notifySplitScreenRequested() {
        val data = JSObject()
        data.put("timestamp", System.currentTimeMillis())
        notifyListeners("splitScreenRequested", data)
    }

    fun notifyFormSelected(form: Map<String, Any>) {
        val data = JSObject()

        form["id"]?.let { data.put("id", it) }
        form["name"]?.let { data.put("name", it) }
        form["links"]?.let { data.put("links", it) }
        form["tenant_id"]?.let { data.put("tenant_id", it) }

        notifyListeners("formSelected", data)
    }
}

File: ios/Plugin/TwilioVideoPlugin.swift
import Foundation
import Capacitor
import TwilioVideo

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(TwilioVideoPlugin)
public class TwilioVideoPlugin: CAPPlugin {

    private static weak var sharedInstance: TwilioVideoPlugin?
    private var videoViewController: VideoCallViewController?

    override public func load() {
        TwilioVideoPlugin.sharedInstance = self
    }

    public static func getInstance() -> TwilioVideoPlugin? {
        return sharedInstance
    }

    @objc func joinRoom(_ call: CAPPluginCall) {
        guard let token = call.getString("token"), !token.isEmpty else {
            call.reject("Token is required")
            return
        }

        let roomName = call.getString("roomName") ?? call.getString("roomId")

        guard let room = roomName, !room.isEmpty else {
            call.reject("Either roomName or roomId must be provided")
            return
        }

        let identity = call.getString("identity")
        let role = call.getString("role")
        let tenantId = call.getInt("tenantId")

        // Validate role if provided
        if let role = role, !["mht", "cct", "patient"].contains(role) {
            call.reject("Invalid role. Allowed values: mht, cct, patient")
            return
        }

        DispatchQueue.main.async {
            let vc = VideoCallViewController()
            vc.accessToken = token
            vc.roomName = room
            vc.userIdentity = identity
            vc.userRole = role
            vc.tenantId = tenantId
            vc.modalPresentationStyle = .fullScreen

            self.bridge?.viewController?.present(vc, animated: true, completion: nil)
            self.videoViewController = vc
        }

        call.resolve()
    }

    @objc func leaveRoom(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            self.videoViewController?.disconnect()
        }
        call.resolve()
    }

    @objc func muteAudio(_ call: CAPPluginCall) {
        let muted = call.getBool("muted") ?? false
        DispatchQueue.main.async {
            self.videoViewController?.muteAudio(muted: muted)
        }
        call.resolve()
    }

    @objc func enableVideo(_ call: CAPPluginCall) {
        let enabled = call.getBool("enabled") ?? true
        DispatchQueue.main.async {
            self.videoViewController?.enableVideo(enabled: enabled)
        }
        call.resolve()
    }

    @objc func flipCamera(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            self.videoViewController?.flipCamera()
        }
        call.resolve()
    }

    @objc func setSpeaker(_ call: CAPPluginCall) {
        let enabled = call.getBool("enabled") ?? true
        DispatchQueue.main.async {
            self.videoViewController?.setSpeaker(enabled: enabled)
        }
        call.resolve()
    }

    @objc func sendUsersList(_ call: CAPPluginCall) {
        print("TwilioVideoPlugin: sendUsersList method invoked")

        guard let selectedRoleKey = call.getString("selectedRoleKey"), !selectedRoleKey.isEmpty else {
            print("TwilioVideoPlugin: sendUsersList failed - selectedRoleKey is required")
            call.reject("selectedRoleKey is required")
            return
        }

        guard ["mht", "cct", "patient", "participant"].contains(selectedRoleKey) else {
            print("TwilioVideoPlugin: sendUsersList failed - invalid selectedRoleKey: \(selectedRoleKey)")
            call.reject("Invalid selectedRoleKey. Allowed values: mht, cct, patient, participant")
            return
        }

        let usersArray = call.getArray("users") as? [[String: Any]] ?? []
        var users: [[String: String]] = []

        print("TwilioVideoPlugin: sendUsersList parsing \(usersArray.count) users")

        for (index, userDict) in usersArray.enumerated() {

            let idValue = userDict["id"]
            let userIdValue = userDict["user_id"]
            let fullName = userDict["full_name"] as? String

            let idStr = (idValue as? String) ?? (idValue as? Int).map { String($0) }
            let userIdStr = (userIdValue as? String) ?? (userIdValue as? Int).map { String($0) }

            if let id = idStr, !id.isEmpty,
               let userId = userIdStr, !userId.isEmpty,
               let name = fullName, !name.isEmpty {

                print("TwilioVideoPlugin: sendUsersList parsed user[\(index)]: \(name) (id=\(id), user_id=\(userId))")

                users.append([
                    "id": id,
                    "user_id": userId,
                    "full_name": name
                ])
            } else {
                print("""
                ⚠️ TwilioVideoPlugin: sendUsersList skipping user[\(index)]
                   raw = \(userDict)
                   idValue = \(String(describing: idValue))
                   userIdValue = \(String(describing: userIdValue))
                   fullName = \(String(describing: fullName))
                """)
            }
        }

        print("TwilioVideoPlugin: sendUsersList calling handleUsersList with \(users.count) users for role: \(selectedRoleKey)")

        DispatchQueue.main.async {
            self.videoViewController?.handleUsersList(selectedRoleKey: selectedRoleKey, users: users)
        }

        call.resolve()
    }

    @objc func sendFormList(_ call: CAPPluginCall) {
        print("TwilioVideoPlugin: sendFormList method invoked")

        let formsArray = call.getArray("forms") as? [[String: Any]] ?? []
        var validatedForms: [[String: Any]] = []

        print("TwilioVideoPlugin: sendFormList parsing \(formsArray.count) forms")

        // Validate and parse only required fields, ignore optional ones safely
        for (index, formDict) in formsArray.enumerated() {
            guard let id = formDict["id"] as? Int,
                  let tenantId = formDict["tenant_id"] as? Int,
                  let name = formDict["name"] as? String, !name.isEmpty,
                  let links = formDict["links"] as? String, !links.isEmpty else {
                print("TwilioVideoPlugin: sendFormList skipping invalid form at index \(index)")
                continue
            }

            // Create validated form with only required fields
            let validatedForm: [String: Any] = [
                "id": id,
                "tenant_id": tenantId,
                "name": name,
                "links": links
            ]

            validatedForms.append(validatedForm)
            print("TwilioVideoPlugin: sendFormList parsed form: \(name) (id=\(id))")
        }

        print("TwilioVideoPlugin: sendFormList calling handleFormsList with \(validatedForms.count) validated forms")

        DispatchQueue.main.async {
            self.videoViewController?.handleFormsList(forms: validatedForms)
        }

        call.resolve()
    }

    // Event notification methods
    public func notifyRoomConnected(roomName: String) {
        notifyListeners("roomConnected", data: ["roomName": roomName])
    }

    public func notifyRoomDisconnected(room: Room) {
        // Build room object structure
        var roomData: [String: Any] = [
            "sid": room.sid,
            "name": room.name,
            "state": roomStateToString(room.state)
        ]

        // Add local participant
        if let localParticipant = room.localParticipant {
            roomData["localParticipant"] = [
                "sid": localParticipant.sid,
                "identity": localParticipant.identity
            ]
        }

        // Add remote participants
        var remoteParticipants: [[String: Any]] = []
        for participant in room.remoteParticipants {
            remoteParticipants.append([
                "sid": participant.sid,
                "identity": participant.identity
            ])
        }
        roomData["remoteParticipants"] = remoteParticipants

        let data: [String: Any] = ["room": roomData]
        notifyListeners("roomDisconnected", data: data)
    }

    private func roomStateToString(_ state: Room.State) -> String {
        switch state {
        case .connecting:
            return "connecting"
        case .connected:
            return "connected"
        case .reconnecting:
            return "reconnecting"
        case .disconnected:
            return "disconnected"
        @unknown default:
            return "unknown"
        }
    }

    public func notifyParticipantJoined(identity: String) {
        notifyListeners("participantJoined", data: ["identity": identity])
    }

    public func notifyParticipantLeft(identity: String) {
        notifyListeners("participantLeft", data: ["identity": identity])
    }

    public func notifyNetworkQualityChanged(identity: String, level: Int, isLocal: Bool) {
        notifyListeners("networkQualityChanged", data: [
            "identity": identity,
            "level": level,
            "isLocal": isLocal
        ])
    }

    public func notifyDominantSpeakerChanged(identity: String?) {
        var data: [String: Any] = [:]
        if let identity = identity {
            data["identity"] = identity
        } else {
            data["identity"] = NSNull()
        }
        notifyListeners("dominantSpeakerChanged", data: data)
    }

    public func notifyRoomAutoClosed(reason: String) {
        notifyListeners("roomAutoClosed", data: ["reason": reason])
    }

    public func notifyRoomError(code: String, message: String, isFatal: Bool) {
        notifyListeners("roomError", data: [
            "code": code,
            "message": message,
            "isFatal": isFatal
        ])
    }

    public func notifyRoleSelected(selectedRoleKey: String, identity: String?, role: String?, tenantId: Int?, roomName: String?, roomSID: String?, secondParticipantRole: String?, secondParticipantIdentity: String?) {
        var data: [String: Any] = ["selectedRoleKey": selectedRoleKey]
        if let identity = identity { data["identity"] = identity }
        if let role = role { data["role"] = role }
        if let tenantId = tenantId { data["tenantId"] = tenantId }
        if let roomName = roomName { data["roomName"] = roomName }
        if let roomSID = roomSID { data["roomSID"] = roomSID }
        if let secondParticipantRole = secondParticipantRole { data["secondParticipantRole"] = secondParticipantRole }
        if let secondParticipantIdentity = secondParticipantIdentity { data["secondParticipantIdentity"] = secondParticipantIdentity }
        notifyListeners("roleSelected", data: data)
    }

    public func notifyUsersListLoaded(selectedRoleKey: String, userCount: Int) {
        notifyListeners("usersListLoaded", data: [
            "selectedRoleKey": selectedRoleKey,
            "userCount": userCount
        ])
    }

    public func notifyUserSelected(id: String, userId: String, fullName: String, selectedRoleKey: String, tenantId: Int?, role: String?, roomName: String?, roomSID: String?) {
        var data: [String: Any] = [
            "id": id,
            "user_id": userId,
            "full_name": fullName,
            "selectedRoleKey": selectedRoleKey
        ]
        if let tenantId = tenantId { data["tenantId"] = tenantId }
        if let role = role { data["role"] = role }
        if let roomName = roomName { data["roomName"] = roomName }
        if let roomSID = roomSID { data["roomSID"] = roomSID }
        notifyListeners("userSelected", data: data)
    }

    public func notifyPopupDismissed(popupType: String, reason: String) {
        notifyListeners("popupDismissed", data: [
            "popupType": popupType,
            "reason": reason
        ])
    }

    public func notifyPopupError(message: String, popupType: String) {
        notifyListeners("popupError", data: [
            "message": message,
            "popupType": popupType
        ])
    }

    public func notifySplitScreenRequested() {
        notifyListeners("splitScreenRequested", data: [
            "timestamp": Date().timeIntervalSince1970 * 1000
        ])
    }

    public func notifyFormSelected(form: [String: Any]) {
        var data: [String: Any] = [:]

        if let id = form["id"] as? Int {
            data["id"] = id
        }
        if let name = form["name"] as? String {
            data["name"] = name
        }
        if let links = form["links"] as? String {
            data["links"] = links
        }
        if let tenantId = form["tenant_id"] as? Int {
            data["tenant_id"] = tenantId
        }

        notifyListeners("formSelected", data: data)
    }
}